File: packages/typescript/src/modules/webhooks.ts
/**
 * WebhooksModule: verify webhook signatures and confirm webhooks via ecollect API.
 */

import { mapReturnCodeToError, WebhookValidationException } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { BASE_URLS } from '../config.js';
import type {
  VerifySessionTokenRequest,
  VerifySessionTokenResponse,
  WebhookPayload,
} from '../types/api.js';
import type { TransactionResult } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';
import { hmacSha256, timingSafeEqual } from '../utils/crypto.js';

export class WebhooksModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
  }

  /**
   * Verify HMAC-SHA256 signature of a webhook payload.
   * The signature is computed over `JSON.stringify(payload)` with `secret`.
   */
  async verifyWebhookSignature(
    payload: Record<string, unknown>,
    signature: string,
    secret: string,
  ): Promise<boolean> {
    const expected = await hmacSha256(JSON.stringify(payload), secret);
    return timingSafeEqual(expected, signature.toLowerCase());
  }

  /**
   * Confirm a webhook by verifying the SessionToken via ecollect API.
   * Uses verifySessionToken endpoint to ensure authenticity.
   */
  async confirmWebhook(
    payload: WebhookPayload,
    sessionToken: string,
  ): Promise<TransactionResult> {
    if (!payload.SessionToken) {
      throw new WebhookValidationException('Webhook payload is missing SessionToken');
    }
    if (!payload.TicketId) {
      throw new WebhookValidationException('Webhook payload is missing TicketId');
    }

    const url = `${BASE_URLS[this.config.environment]}/verifySessionToken`;
    const activeToken = sessionToken || (await this.session.getActive());

    const body: VerifySessionTokenRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: activeToken,
      SessionTokenToVerify: payload.SessionToken,
      TicketIdToVerify: payload.TicketId,
    };

    const res = await this.http.post<VerifySessionTokenRequest, VerifySessionTokenResponse>(
      url,
      body,
    );

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode, `verifySessionToken failed: ${res.ReturnCode}`);
    }

    // Return the transaction data from the webhook payload
    return {
      returnCode: payload.ReturnCode,
      ticketId: payload.TicketId,
      tranState: payload.TranState as TransactionResult['tranState'],
      trazabilityCode: payload.TrazabilityCode,
      transValue: payload.TransValue,
      transVatValue: payload.TransVatValue,
      payCurrency: payload.PayCurrency,
      bankProcessDate: payload.BankProcessDate,
      fiCode: payload.FICode,
      fiName: payload.FiName,
      paymentSystem: payload.PaymentSystem,
      referenceArray: payload.ReferenceArray,
      srvCode: payload.SrvCode,
    };
  }

  /**
   * Build the response body that ecollect expects from a webhook endpoint.
   */
  static buildWebhookResponse(success: boolean): { ReturnCode: string } {
    return { ReturnCode: success ? 'SUCCESS' : 'FAIL_SYSTEM' };
  }
}

File: packages/python/ecollect/modules/webhooks.py
"""Webhook verification and confirmation module."""
import logging
from typing import Any, Dict, Optional, Union

from ecollect.config import EcollectConfig
from ecollect.exceptions import WebhookValidationError, raise_for_return_code
from ecollect.utils.crypto import compute_hmac_sha256, payload_to_bytes, verify_hmac_sha256
from ecollect.utils.http import AsyncHttpClient, run_sync

logger = logging.getLogger(__name__)


class WebhooksModule:
    """Handles webhook signature verification and session token confirmation."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session

    def verify_webhook_signature(
        self,
        payload: Union[str, bytes, Dict[str, Any]],
        secret: str,
        signature: str,
    ) -> bool:
        """Verify an HMAC-SHA256 webhook signature.

        Args:
            payload: Raw request body (str/bytes) or dict that will be
                     serialised to canonical JSON.
            secret: Shared secret known to both ecollect and the merchant.
            signature: Signature provided in the webhook request header.

        Returns:
            True if the signature is valid.

        Raises:
            WebhookValidationError: If the signature does not match.
        """
        raw = payload_to_bytes(payload)
        if not verify_hmac_sha256(raw, secret, signature):
            raise WebhookValidationError("Webhook HMAC-SHA256 signature mismatch")
        return True

    async def confirm_webhook(
        self,
        session_token_to_verify: str,
        ticket_id: Optional[int] = None,
    ) -> bool:
        """Confirm webhook authenticity via the verifySessionToken API.

        Args:
            session_token_to_verify: The SessionToken field received in the webhook payload.
            ticket_id: Optional TicketId to verify against the token.

        Returns:
            True if ecollect confirms the token is valid.

        Raises:
            WebhookValidationError: If verification fails.
        """
        current_token = await self._session.get_active()
        url = self._config.endpoint_url("verifySessionToken")
        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": current_token,
            "SessionTokenToVerify": session_token_to_verify,
        }
        if ticket_id is not None:
            payload["TicketIdToVerify"] = ticket_id

        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        raise_for_return_code(return_code, str(data))
        return True

    # ------------------------------------------------------------------
    # Sync wrappers
    # ------------------------------------------------------------------

    def confirm_webhook_sync(
        self,
        session_token_to_verify: str,
        ticket_id: Optional[int] = None,
    ) -> bool:
        return run_sync(self.confirm_webhook(session_token_to_verify, ticket_id))
